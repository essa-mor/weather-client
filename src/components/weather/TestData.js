export const data = [
	{ dt: 1535274000, day: '2018-08-23 12:00:00', maxTemperature: 29.39, minTemperature: 27.66, img: '01d' },
	{ dt: 1535275000, day: '2018-08-24 12:00:01', maxTemperature: 30.39, minTemperature: 27.14, img: '01d' },
	{ dt: 1535276000, day: '2018-08-25 12:00:02', maxTemperature: 31.39, minTemperature: 26.95, img: '01d' },
	{ dt: 1535277000, day: '2018-08-26 12:00:03', maxTemperature: 32.39, minTemperature: 27.01, img: '02d' },
	{ dt: 1535278000, day: '2018-08-27 12:00:04', maxTemperature: 33.69, minTemperature: 26.9, img: '03d' }];