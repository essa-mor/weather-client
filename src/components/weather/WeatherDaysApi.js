import mem from 'mem';

export function customFetch(url) {
	return fetch(url).then(response => response.json());
}

export function getMemCustomFetch(func, maxAge) {
	return mem(func, { maxAge });
}
const maxAge = 10 * 60 * 1000;
const memCustomFetch = getMemCustomFetch(customFetch, maxAge);

export function getWeatherDays() {
	return memCustomFetch('https://api.openweathermap.org/data/2.5/forecast/?q=Haifa,il&units=metric&APPID=8c6cf469699d615e37361375f85fb917')
		.then(data => getDailyWeather(data));
}

export function getWeatherDay(dt) {
	return memCustomFetch('https://api.openweathermap.org/data/2.5/forecast/?q=Haifa,il&units=metric&APPID=8c6cf469699d615e37361375f85fb917')
		.then(data => extractDayWeather(data, dt));
}

export function extractDayWeather(data, dt) {
	const result = [];
	const { list } = data;
	const firstIndex = list.findIndex(r => r.dt >= dt);
	for (let i = firstIndex; i < list.length && result.length < 8; i++) {
		const { dt_txt, dt } = list[i];
		const { temp } = list[i].main;
		result.push({ dt, dt_txt, temp });
	}
	return result;
}

export function getDailyWeather(data) {
	const result = [];
	const { list } = data;
	let currDtTxt = null;
	for (let i = 0; i < list.length; i++) {
		const { dt_txt, dt } = list[i];
		const { temp_max, temp_min } = list[i].main;
		const { icon } = list[i].weather[0];
		if (currDtTxt !== dt_txt.split(' ')[0]) {
			result.push({ dt, day: dt_txt, maxTemperature: temp_max, minTemperature: temp_min, img: icon });
			currDtTxt = dt_txt.split(' ')[0];
		}
	}
	return result;
}