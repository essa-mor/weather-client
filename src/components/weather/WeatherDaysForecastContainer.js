import React from 'react';
import { withRouter } from 'react-router-dom';

import WeatherDaysForecast from './WeatherDaysForecast';
import { getWeatherDays } from './WeatherDaysApi';

export class WeatherDaysForecastContainer extends React.PureComponent {
	constructor(props) {
		super(props);
		this.state = { data: [], selectedDt: null };
		this.handleOnChange = this.handleOnChange.bind(this);
	}

	handleOnChange(dt) {
		const { history } = this.props;
		history.push(`/${dt}`);
	}

	componentDidMount() {
		const { location } = this.props;
		getWeatherDays().then(data => this.setState({ data }));
		const params = location && location.pathname.split('/').filter(i => i !== '');
		if (params != null && params.length > 0) {
			this.setState({ selectedDt: parseInt(params[0]) });
		}
	}

	render() {
		const { data, selectedDt } = this.state;
		return <WeatherDaysForecast onChange={this.handleOnChange} data={data} selectedDt={selectedDt} />;
	}
}

export default withRouter(WeatherDaysForecastContainer);