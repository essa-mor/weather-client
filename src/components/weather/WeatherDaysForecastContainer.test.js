import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { WeatherDaysForecastContainer } from './WeatherDaysForecastContainer';
import { getWeatherDays } from './WeatherDaysApi';
import { data } from './TestData';

jest.mock('./WeatherDaysApi');

let wrapper;
describe('WeatherDaysForecastContainer test', () => {
	it('renders correctly - render all', async () => {
		getWeatherDays.mockImplementation(() => Promise.resolve(data));
		const tree = mount(<WeatherDaysForecastContainer history={[]} />);
		await tree.instance().componentDidMount();
		tree.update();
		expect(toJson(tree)).toMatchSnapshot();
	});

	it('renders correctly', done => {
		getWeatherDays.mockImplementation(() => Promise.resolve(data));
		wrapper = mount(<WeatherDaysForecastContainer history={[]} />);
		setTimeout(() => {
			wrapper.update();
			expect(wrapper.find('.weather_day_indicator')).toHaveLength(5);
			done();
		});
	});

	it('Select first item', async () => {
		getWeatherDays.mockImplementation(() => Promise.resolve(data));
		wrapper = mount(<WeatherDaysForecastContainer history={[]} />);
		await wrapper.instance().componentDidMount();
		wrapper.update();
		expect(wrapper.state().selectedDt).toBe(1535274000);
	});

	it('Select from history', () => {
		getWeatherDays.mockImplementation(() => Promise.resolve(data));
		wrapper = mount(<WeatherDaysForecastContainer history={[]} location={{ pathname: '1535275000' }} />);
		expect(wrapper.state().selectedDt).toBe(1535275000);
	});
});