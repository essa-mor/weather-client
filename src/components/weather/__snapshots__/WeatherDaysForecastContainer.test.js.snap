// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherDaysForecastContainer test renders correctly - await promise resolve 1`] = `
<WeatherDaysForecastContainer
  history={
    Array [
      "/weather-client/1535274000",
    ]
  }
>
  <WeatherDaysForecast
    data={
      Array [
        Object {
          "day": "2018-08-23 12:00:00",
          "dt": 1535274000,
          "img": "01d",
          "maxTemperature": 29.39,
          "minTemperature": 27.66,
        },
        Object {
          "day": "2018-08-24 12:00:01",
          "dt": 1535275000,
          "img": "01d",
          "maxTemperature": 30.39,
          "minTemperature": 27.14,
        },
        Object {
          "day": "2018-08-25 12:00:02",
          "dt": 1535276000,
          "img": "01d",
          "maxTemperature": 31.39,
          "minTemperature": 26.95,
        },
        Object {
          "day": "2018-08-26 12:00:03",
          "dt": 1535277000,
          "img": "02d",
          "maxTemperature": 32.39,
          "minTemperature": 27.01,
        },
        Object {
          "day": "2018-08-27 12:00:04",
          "dt": 1535278000,
          "img": "03d",
          "maxTemperature": 33.69,
          "minTemperature": 26.9,
        },
      ]
    }
    onChange={[Function]}
    selectedDt={1535274000}
  >
    <div
      className="weather_days_forecast"
    >
      <WeatherDayIndicator
        day="2018-08-23 12:00:00"
        dt={1535274000}
        img="http://openweathermap.org/img/w/01d.png"
        isSelected={true}
        key="2018-08-23 12:00:00"
        maxTemperature={29.39}
        minTemperature={27.66}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator weather_day_indicator_selected"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Thu
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/01d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              29
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              28
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
      <WeatherDayIndicator
        day="2018-08-24 12:00:01"
        dt={1535275000}
        img="http://openweathermap.org/img/w/01d.png"
        isSelected={false}
        key="2018-08-24 12:00:01"
        maxTemperature={30.39}
        minTemperature={27.14}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Fri
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/01d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              30
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              27
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
      <WeatherDayIndicator
        day="2018-08-25 12:00:02"
        dt={1535276000}
        img="http://openweathermap.org/img/w/01d.png"
        isSelected={false}
        key="2018-08-25 12:00:02"
        maxTemperature={31.39}
        minTemperature={26.95}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Sat
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/01d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              31
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              27
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
      <WeatherDayIndicator
        day="2018-08-26 12:00:03"
        dt={1535277000}
        img="http://openweathermap.org/img/w/02d.png"
        isSelected={false}
        key="2018-08-26 12:00:03"
        maxTemperature={32.39}
        minTemperature={27.01}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Sun
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/02d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              32
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              27
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
      <WeatherDayIndicator
        day="2018-08-27 12:00:04"
        dt={1535278000}
        img="http://openweathermap.org/img/w/03d.png"
        isSelected={false}
        key="2018-08-27 12:00:04"
        maxTemperature={33.69}
        minTemperature={26.9}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Mon
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/03d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              34
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              27
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
    </div>
  </WeatherDaysForecast>
</WeatherDaysForecastContainer>
`;

exports[`WeatherDaysForecastContainer test renders correctly - call componentDidMount 1`] = `
<WeatherDaysForecastContainer
  history={
    Array [
      "/weather-client/1535274000",
    ]
  }
>
  <WeatherDaysForecast
    data={
      Array [
        Object {
          "day": "2018-08-23 12:00:00",
          "dt": 1535274000,
          "img": "01d",
          "maxTemperature": 29.39,
          "minTemperature": 27.66,
        },
        Object {
          "day": "2018-08-24 12:00:01",
          "dt": 1535275000,
          "img": "01d",
          "maxTemperature": 30.39,
          "minTemperature": 27.14,
        },
        Object {
          "day": "2018-08-25 12:00:02",
          "dt": 1535276000,
          "img": "01d",
          "maxTemperature": 31.39,
          "minTemperature": 26.95,
        },
        Object {
          "day": "2018-08-26 12:00:03",
          "dt": 1535277000,
          "img": "02d",
          "maxTemperature": 32.39,
          "minTemperature": 27.01,
        },
        Object {
          "day": "2018-08-27 12:00:04",
          "dt": 1535278000,
          "img": "03d",
          "maxTemperature": 33.69,
          "minTemperature": 26.9,
        },
      ]
    }
    onChange={[Function]}
    selectedDt={1535274000}
  >
    <div
      className="weather_days_forecast"
    >
      <WeatherDayIndicator
        day="2018-08-23 12:00:00"
        dt={1535274000}
        img="http://openweathermap.org/img/w/01d.png"
        isSelected={true}
        key="2018-08-23 12:00:00"
        maxTemperature={29.39}
        minTemperature={27.66}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator weather_day_indicator_selected"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Thu
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/01d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              29
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              28
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
      <WeatherDayIndicator
        day="2018-08-24 12:00:01"
        dt={1535275000}
        img="http://openweathermap.org/img/w/01d.png"
        isSelected={false}
        key="2018-08-24 12:00:01"
        maxTemperature={30.39}
        minTemperature={27.14}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Fri
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/01d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              30
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              27
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
      <WeatherDayIndicator
        day="2018-08-25 12:00:02"
        dt={1535276000}
        img="http://openweathermap.org/img/w/01d.png"
        isSelected={false}
        key="2018-08-25 12:00:02"
        maxTemperature={31.39}
        minTemperature={26.95}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Sat
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/01d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              31
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              27
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
      <WeatherDayIndicator
        day="2018-08-26 12:00:03"
        dt={1535277000}
        img="http://openweathermap.org/img/w/02d.png"
        isSelected={false}
        key="2018-08-26 12:00:03"
        maxTemperature={32.39}
        minTemperature={27.01}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Sun
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/02d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              32
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              27
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
      <WeatherDayIndicator
        day="2018-08-27 12:00:04"
        dt={1535278000}
        img="http://openweathermap.org/img/w/03d.png"
        isSelected={false}
        key="2018-08-27 12:00:04"
        maxTemperature={33.69}
        minTemperature={26.9}
        onClick={[Function]}
      >
        <div
          className="weather_day_indicator"
          onClick={[Function]}
        >
          <span
            className="weather_day"
          >
            Mon
          </span>
          <img
            alt=""
            src="http://openweathermap.org/img/w/03d.png"
          />
          <div
            className="weather_temperatures"
          >
            <span
              className="weather_max_temperature"
            >
              34
              °
            </span>
            <span
              className="weather_min_temperature"
            >
              27
              °
            </span>
          </div>
        </div>
      </WeatherDayIndicator>
    </div>
  </WeatherDaysForecast>
</WeatherDaysForecastContainer>
`;
